Refacto plugin loader (PXY-482)

fteychene opened the pull request:
# What does this PR do ?

Add abstraction for plugin load
Add default plugin loader based on java ServiceLoader

Remove all reflection code from plugin loading
Change plugin loading in interceptorpool service

ehumber commented on logger-interceptor/src/main/java/io/conduktor/example/loggerinterceptor/AllLoggerInterceptor.java:
```diff
@@ -36,4 +36,9 @@ public CompletionStage<AbstractRequestResponse> intercept(AbstractRequestRespons
+
+    @Override
+    public Class<AbstractRequestResponse> type() {
+        return AbstractRequestResponse.class;
```
Seems a shame to make people implement this, it's boiler plate.  Is there any way to get type() without the interceptor implementor having to write this? (I think not, because it's not possible in Java, but asking in case you can think of a way).

ehumber replied:
Also, if they have to type the Type, then this is somewhere people can make mistakes

fteychene replied:
We need some reified system to do that. I have some ideas in other languages or by having an annotation processor at compile time but except that I don't have any other solution :disappointed:

fteychene replied:
I added an indirection layer with `InterceptorProvider` for now to keep Interceptor easy and let this on Plugin site.
I'm also searching for alternatives ways

ehumber commented on interceptor-framework/src/main/java/io/conduktor/gateway/interceptor/Plugin.java:
```diff
@@ -22,30 +22,19 @@
-    void configure(Map<String,Object> config) throws InterceptorConfigurationException;
-
-    List<Interceptor> getInterceptors();
+    default String pluginId() {
```
Is this what will go in the application.yaml?  It's still the Plugin class name.  Is there any way we can swap it to be the Interceptor name instead?  

I guess swap this so it returns "com.my.company.Interceptor" as a String?

fteychene replied:
It's the idea to use this in interceptor config (even if I didn't code it ATM)

Interceptor name ? A plugin can have multiple interceptors so which one ?
that's also why I defined it as default. To let plugin maintainer be able to define it but without breaking the current interface implementation

ehumber commented on logger-interceptor/src/main/java/io/conduktor/example/loggerinterceptor/LoggerInterceptorPlugin.java:
```diff
@@ -26,23 +27,22 @@
                 new ResponseLoggerInterceptor());
     }
+
+    @Override
```
If Plugin is an abstract class not an interface, they won't need to implement pluginId any more.  Again it seems like boilerplate code that will always be the same

fteychene replied:
Yes I also thought about abstract to test if it could help us.
Wanted to check in this direction

fteychene replied:
BTW th eabstrat is for other cases. For this one it's just a mistake. It don't need to be defined

ehumber commented:
I know this is a prototype and is looking for feedback.  As part of that I think it would be beneficial to update the readme.md, the run-gateway.sh script and and the write-your-own-interceptor.md files.

It might not be the final version, but writing up how a user is going to actually use this code will help highlight what we've not thought of, and things we won't spot with just a code review.

fteychene commented:
I changed the type definition for interceptor for a new `InterceptorProvider` that is responsible to provide type for each Interceptor and updated the doc

ehumber commented on README.md:
```diff
@@ -134,8 +134,10 @@ bootstrap.servers=localhost:9092
 kafkaSelector:
-  file:
-    path:  - the path to the file containing Kafka connection properties e.g. config/kafka.config
+  type: file|env
```
Do we want people using env here? I left it at just file on purpose to keep config simple.


I think prefix needs a bit more of an explanation, an example would help.

Also, the conduktor-docs need updating to match.

fteychene replied:
Using the docker image it's way easier to use the env config.

fteychene replied:
Other question is : Why do we ocument this here and in conduktor-docs ?
We should only document config once and reference it across other docs

ehumber replied:
Agreed - I'll fix that :)

ehumber commented on gateway-core/config/kafka.config:
```diff
@@ -1,3 +1,3 @@
-bootstrap.servers=localhost:9092
+bootstrap.servers=localhost:29092
```
9092 was on purpose.  This is the default port when you follow the AK quick start.  If we change it to 29092 that matches our docker config, but is a less smooth experience for a new user.  I think it should stay at 9092.

fteychene replied:
Will come back to 9092 it should not be be there

fteychene replied:
Ok got it. It's cause docker-compose is exposing to 29092. So I'm gonna change this to be aligned cause I don't like to have a port on default config that don't work with embeded dev env

thomaskwscott reviewed (approved):
2 small things but otherwise LGTM

thomaskwscott commented on interceptor-framework/README.md:
```diff
@@ -30,14 +30,17 @@ For example, a logging Interceptor may include two `Interceptor.java` implementa
 It also processes the interceptor configuration, which is specified in `application.yaml`.
 
-### Interceptor jar
+### Interceptor framework jar
```
no need, only a nit
